File: General Testing/Controlsystem/Controlsystem.py
import time
import lgpio
import cv2
# import PWM_signals

#initializeCamera()
# global PWM_signal_steering
# global PWM_signal_throttle
# V_STEADY = 20.5     # Max velocity (m/s)
# ACCEL = 10.25    # Max acceleration (m/s^2)
# D_TOTAL   = 82.0   # Total displacement (m)

# PWM_signal_throttle = 15
# PWM_signal_steering = 15
#                                         #Kp         Ki         Kd  
# steeringController = SteeringController(0.061388 , 0.682021 , 0.0)
# throttleController = ThrottleController(1.0 , 1.0 , 1.0)

# Initialize camera
cap = cv2.VideoCapture(0)

def runAndPrintControlSystem(HANDLE, frequency, steeringController, throttleController, velocityProfile, desiredDistToCenterline):

        start_time = time.time()
        print(f"\nTime: {start_time}:")


        throttleController.update_pwm(start_time, velocityProfile)
        steeringController.update_pwm(desiredDistToCenterline)
                
        # Commenting to combine with cameraAndThrottle.py
        '''
        PWM_signal_throttle = throttleController.pwm
        PWM_signal_steering = steeringController.pwm
        print(f"Desired Position: {throttleController.xd:.2f} m") 
        print(f"Current Position: {throttleController.x:.2f} m")
        print(f"Desired Velocity: {throttleController.v:.2f} m/s") 
        print(f"Current Velocity: {throttleController.x:.2f} m")
        print()  # Prints a blank line
        print(f"Desired Distance to Centerline: {steeringController.desired_dist_to_centerline:.2f} m")
        print(f"Current Steering Angle: {steeringController.current_dist_to_centerline:.2f} m")
        print(f"PWM throttle: {throttleController.pwm:.2f} %")
        print(f"PWM steering: {steeringController.pwm:.2f} %")
        print(f"End Time: {time.time()}")


        # Generate PWM
        lgpio.tx_pwm(HANDLE, 13, frequency, PWM_signal_throttle)
        lgpio.tx_pwm(HANDLE, 12, frequency, PWM_signal_steering)
        '''

File: General Testing/Controlsystem/Controlsystem.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <unistd.h> // For usleep function

#define V_STEADY 20.5  // Max velocity (m/s)
#define ACCEL 10.25     // Max acceleration (m/s^2)
#define D_TOTAL 82.0    // Total displacement (m)

double PWM_signal_throttle = 15;
double PWM_signal_steering = 15;

typedef struct {
    double Kp, Ki, Kd;
    double pwm;
    double xd, vd;
    double x, x_previous;
    double v;
} ThrottleController;

typedef struct {
    double kp, ki, kd;
    double pwm;
    double desired_steering_angle;
    double current_steering_angle;
} SteeringController;

void main() {
    ThrottleController throttleController = {1.0, 1.0, 1.0, 0, 0, 0, 0, 0};
    SteeringController steeringController = {0.061388, 0.682021, 0.0, 15, 15};
    
    double start_time = time(NULL);
    
    while (throttleController.x <= D_TOTAL) {
        PWM_signal_throttle = throttleController.pwm;
        PWM_signal_steering = steeringController.pwm;
        
        printf("Desired Position: %.2f m\n", throttleController.xd);
        printf("Current Position: %.2f m\n", throttleController.x);
        printf("Desired Velocity: %.2f m/s\n", throttleController.vd);
        printf("Current Velocity: %.2f m/s\n", throttleController.v);
        printf("\n"); // Blank line
        printf("Desired Steering Angle: %.2f degrees\n", steeringController.desired_steering_angle);
        printf("Current Steering Angle: %.2f degrees\n", steeringController.current_steering_angle);
        
        usleep(100000); // 100 ms delay
    }
}
